src/statsd.c

```diff
@@ -0,0 +1,190 @@
+#ifndef _TEST_TARGET
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <chan/chan.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <signal.h>
+#include <pcp/pmapi.h>
+
+#include "statsd.h"
+#include "config-reader.h"
+#include "network-listener.h"
+#include "aggregators.h"
+#include "aggregator-metrics.h"
+#include "aggregator-stats.h"
+#include "pmda-callbacks.h"
+#include "utils.h"
+#include "../domain.h"
+
+void signal_handler(int num) {
+    if (num == SIGUSR1) {
+        aggregator_request_output();
+    }
+}
+
+/**
+ * Initializes structure which is used as private data container accross all PCP related callbacks
+ * @arg args - All args passed to 'PCP exchange' thread
+ */
+static void
+init_data_ext(
+    struct pmda_data_extension* data,
+    struct agent_config* config,
+    struct pmda_metrics_container* metrics_storage,
+    struct pmda_stats_container* stats_storage
+) {
+    data->config = config;
+    data->pcp_metric_count = 0;
+    data->pcp_instance_domain_count = 0;
+    data->metrics_storage = metrics_storage;
+    data->stats_storage = stats_storage;
+    data->reload = 0;
+    data->notify = 0;
+}
+
+/**
+ * Registers hardcoded metrics that are registered before PMDA agent initializes itself fully
+ * @arg pi - pmdaInterface carries all PCP stuff
+ * @arg data - 
+ */
+static void
+create_statsd_hardcoded_metrics(pmdaInterface* pi, struct pmda_data_extension* data) {
+    size_t i;
+    size_t hardcoded_count = 6;
+    data->pcp_metrics = malloc(hardcoded_count * sizeof(pmdaMetric));
+    ALLOC_CHECK("Unable to allocate space for static PMDA metrics.");
+    for (i = 0; i < hardcoded_count; i++) {
+        data->pcp_metrics[i].m_user = data;
+        data->pcp_metrics[i].m_desc.pmid = pmID_build(pi->domain, 0, i);
+        data->pcp_metrics[i].m_desc.type = PM_TYPE_U64;
+        data->pcp_metrics[i].m_desc.indom = PM_INDOM_NULL;
+        data->pcp_metrics[i].m_desc.sem = PM_SEM_INSTANT;
+        if (i < 4) {
+            memset(&data->pcp_metrics[i].m_desc.units, 0, sizeof(pmUnits));
+        } else {
+            // time_spent_parsing / time_spent_aggregating
+            data->pcp_metrics[i].m_desc.units.dimSpace = 0;
+            data->pcp_metrics[i].m_desc.units.dimTime = 0;
+            data->pcp_metrics[i].m_desc.units.dimCount = 0;
+            data->pcp_metrics[i].m_desc.units.pad = 0;
+            data->pcp_metrics[i].m_desc.units.scaleSpace = 0;
+            data->pcp_metrics[i].m_desc.units.scaleTime = PM_TIME_NSEC;
```

Comment:
In the help you say microseconds and this is NSEC which reads like "nanoseconds". I haven't check in the manual, but this must match otherwise PCP will convert it incorrectly.

Reply:
Fixed